Microsoft.Maui.Devices.Flashlight: Check if supported

vividos opened the pull request:
### Description of Change

This adds the `CheckIsSupported()` call to the `Microsoft.Maui.Devices.Flashlight` API, as discussed in #13458.

### Issues Fixed

Fixes #13458

@jfversluis Feel free to change the name of the added method. In #13458 we didn't have a clear consent what name to use. Thanks!

vividos committed:
implemented checking if Flashlight is supported on the device

vividos committed:
added initializing IsSupported flag using CheckIsSupported()

jfversluis commented:
/azp run

rmarinho reviewed (changes requested):
```
D:\a\_work\1\s\src\Essentials\src\Flashlight\Flashlight.ios.cs(16,11): error CS0029: Cannot implicitly convert type 'bool' to 'System.Threading.Tasks.Task<bool>' [D:\a\_work\1\s\src\Essentials\src\Essentials.csproj::TargetFramework=net7.0-ios]
D:\a\_work\1\s\src\Essentials\src\Flashlight\Flashlight.ios.cs(16,11): error CS0029: Cannot implicitly convert type 'bool' to 'System.Threading.Tasks.Task<bool>' [D:\a\_work\1\s\src\Essentials\src\Essentials.csproj::TargetFramework=net7.0-maccatalyst]
    69 Warning(s)
    2 Error(s)
```

rmarinho commented on src/Essentials/samples/Samples/ViewModel/FlashlightViewModel.cs:
```diff
@@ -14,6 +14,7 @@ public class FlashlightViewModel : BaseViewModel
 		public FlashlightViewModel()
 		{
 			ToggleCommand = new Command(OnToggle);
+			Device.BeginInvokeOnMainThread(this.InitViewModel);
```
drop this.

vividos replied:
What do you suggest instead to initialize the IsSupported property?

vividos replied:
Ah sorry I misunderstood, you meant remove `this.`

rmarinho commented on src/Essentials/samples/Samples/ViewModel/FlashlightViewModel.cs:
```diff
@@ -30,6 +31,11 @@ public bool IsSupported
 			set => SetProperty(ref isSupported, value);
 		}
 
+		private async Task InitViewModel()
```
drop private keyboard since it's the default

vividos replied:
I'll remove it, thanks!

mandel-macaque commented on src/Essentials/src/Flashlight/Flashlight.ios.cs:
```diff
@@ -6,6 +6,17 @@ namespace Microsoft.Maui.Devices
+		public Task<bool> CheckIsSupported()
+		{
+			var captureDevice = AVCaptureDevice.GetDefaultDevice(AVMediaTypes.Video);
+			return captureDevice != null &&
```
We are trying to move away from == null or != null
```suggestion
			return captureDevice is not null &&
```

mandel-macaque commented on src/Essentials/src/Flashlight/Flashlight.ios.cs:
```diff
@@ -6,6 +6,17 @@ namespace Microsoft.Maui.Devices
+		/// <returns><see langword="true"/> when the flashlight is available, or <see langword="false"/> when not</returns>
+		public Task<bool> CheckIsSupported()
+		{
+			var captureDevice = AVCaptureDevice.GetDefaultDevice(AVMediaTypes.Video);
```
This is me asking questions. We are getting the default video device, but we are not saying the position.

There [is an overload of this method ](https://learn.microsoft.com/en-us/dotnet/api/avfoundation.avcapturedevice.getdefaultdevice?view=xamarin-ios-sdk-12#avfoundation-avcapturedevice-getdefaultdevice(avfoundation-avcapturedevicetype-system-string-avfoundation-avcapturedeviceposition)) that specifies the position of the device (front or back). Can we use the front camera flash as a flashlight? If we cannot, we should specify the position directly to ensure we are using the correct device.

vividos replied:
The code was copied from `Toggle()`, below. I'll check what the default parameter is. I think most devices on iOS only have a back side flashlight, so I guess that would be the better side to check.

vividos replied:
I reviewed the two `AVCaptureDevice.GetDefaultDevice()` methods. I would keep te same call as in the `Toggle()` call, below in the same source file, when switching on/off the flashlight. If you want to change what `GetDefaultDevice()` call to use, in order to give the caller the possibility to choose which flashlight to switch, feel free to do so. I think this PR is not the proper context for that, though. I also don't think I saw any iOS devices yet that had a front camera flashlight (but once had an iPad mini that had no back flashlight).

mandel-macaque commented on src/Essentials/src/Flashlight/Flashlight.uwp.cs:
```diff
@@ -48,6 +48,20 @@ async Task FindLampAsync()
+
+			lock (locker)
+			{
+				return hasLoadedLamp && lamp != null;
```
```suggestion
				return hasLoadedLamp && lamp is no null;
```

hartez replied:
```suggestion
				return hasLoadedLamp && lamp is not null;
```

vividos commented:
Thanks for all the review comments, I'll apply the changes on next Saturday

mattleibow commented on src/Essentials/src/Flashlight/Flashlight.shared.cs:
```diff
@@ -8,6 +8,12 @@ namespace Microsoft.Maui.Devices
+		/// Checks if the flashlight is available and can be turned on or off.
+		/// </summary>
+		/// <returns><see langword="true"/> when the flashlight is available, or <see langword="false"/> when not</returns>
+		Task<bool> CheckIsSupported();
```
Should this have the `Async` suffix to indicate that it needs to be awaited?

vividos replied:
What is your policy regarding to `async` methods?

mattleibow replied:
I think we typically have Async suffix - especially since essentials has a mix.

vividos replied:
Ok I'll add the suffix.

jfversluis commented:
Take your time @vividos ! Thank you for this! And thanks everyone for the input 😄

vividos committed:
fix compiling under iOS

vividos committed:
renamed CheckIsSupported() to CheckIsSupportedAsync()

vividos committed:
removed .this and private from sample code and added missing using

vividos committed:
replaced != null with is not null

vividos commented:
I fixed the iOS compile error (I changed the TargetFrameworks in Essentials.csproj a bit for compiling and was missing iOS errors) and also did the other requested changes. The PR should be ready now.

rmarinho commented:
/azp run

vividos committed:
trying to fix build by replacing Device.BeginInvokeOnMainThread()

rmarinho commented:
/azp run

vividos commented on src/Essentials/src/Flashlight/Flashlight.android.cs:
```diff
@@ -34,6 +34,12 @@ internal bool IsSupported
+		/// Checks if the flashlight is available and can be turned on or off.
+		/// </summary>
+		/// <returns><see langword="true"/> when the flashlight is available, or <see langword="false"/> when not</returns>
+		public Task<bool> CheckIsSupportedAsync() => Task.FromResult(IsSupported);
```
We already discussed this in #13458 and didn't come to a consent. So please discuss the name and I'll change that afterwards.

mattleibow replied:
IsSupportedAsync sounds good too!

vividos commented:
@jfversluis How do we proceed with this PR? Someone has to decide how to name that method...

vividos commented:
No problem, I'll rename the method.

vividos committed:
renamed CheckIsSupportedAsync() to IsSupportedAsync()

jfversluis commented:
Thank you for another wonderful PR @vividos !

vividos commented:
You're welcome ☺️
